Sources/ForageSDK/Foundation/Network/Collector.swift

```diff
@@ -323,6 +337,157 @@ class BasisTheoryWrapper: VaultCollector {
+    }
+    
+    func sendData<T: Decodable>(path: String, vaultAction: VaultAction, extraData: [String : Any], completion: @escaping (T?, ForageError?) -> Void) {
+        let semaphore = DispatchSemaphore(value: 0)
+        var pin = ""
+        
+        DispatchQueue.main.async {
+            guard let textElementValue = self.textElement.text else {
+                return completion(nil, CommonErrors.INCOMPLETE_PIN_ERROR)
+            }
+            pin = textElementValue
+            semaphore.signal()
+        }
+
+        // waiting for the value of the text field to be read from the main queue and saved in `pin`
+        semaphore.wait()
+
+        var body: [String: String] = ["pin": pin]
+
+        for (key, value) in extraData {
+            if key == tokenKey, let paymentMethodToken = value as? String {
+                do {
+                    let token = try getPaymentMethodToken(paymentMethodToken: paymentMethodToken)
+                    body[key] = token
+                } catch {
+                    logger?.critical(
+                        "Failed to send data to Rosetta proxy. Rosetta token not found on card",
+                        error: error,
+                        attributes: nil
+                    )
+                    return completion(nil, CommonErrors.UNKNOWN_SERVER_ERROR)
+                }
+            }
+        }
+        
+        let measurement = VaultProxyResponseMonitor.newMeasurement(vault: VaultType.forage, action: vaultAction)
+            .setPath(path)
+            .setMethod(.post)
+        
+        measurement.start()
+        
+        let url = URL(string: "https://\(forageVaultConfig.vaultBaseURL)/proxy\(path)")!
+        var request = URLRequest(url: url)
+        request.httpMethod = "POST"
+        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
+        for header in customHeaders {
+            if header.key == "Session-Token" {
+                request.setValue(header.value, forHTTPHeaderField: "Authorization")
+            } else {
+                request.setValue(header.value, forHTTPHeaderField: header.key)
+            }
+        }
+        request.httpBody = try! JSONSerialization.data(withJSONObject: body)
+        
+        let task = URLSession.shared.dataTask(with: request) { data, response, error in
+            self.handleResponse(response: response, data: data, error: error, measurement: measurement, completion: completion)
+        }
+        
+        task.resume()
+    }
+    
+    func handleResponse<T: Decodable>(response: URLResponse?, data: Data?, error: Error?, measurement: NetworkMonitor, completion: (T?, ForageError?) -> Void) {
+        measurement.end()
+        
+        let code = (response as? HTTPURLResponse)?.statusCode ?? 500
+        
+        measurement.setHttpStatusCode(code).logResult()
+
+        // If an error is explicitly returned from Rosetta, log the error and return
+        if let error = error {
```

Comment:
This kind of error may be returned from the Rosetta proxy much more often than from the external proxies. This line is basically unreachable for BT and VGS, but I wonder if we should downgrade the logging of this error to non-critical because this is where normal errors like 429 would appear. Disregard if that's a dumb question and you already know the answer :)

Reply:
Maybe just `.error` instead of `.critical`?

Reply:
[799b5d0](https://github.com/teamforage/forage-ios-sdk/pull/169/commits/799b5d07b8563b1ba67ccc8b9589dc7501c2b383)